src/pip/_internal/resolution/resolvelib/providers.py

```diff
@@ -0,0 +1,66 @@
+import operator
+
+from pip._vendor.packaging.requirements import (
+    Requirement as PEP440Requirement,
+)
+from pip._vendor.packaging.utils import canonicalize_name
+
+from pip._internal.resolution.resolvelib.models import (
+    DirectCandidate,
+    ExtrasCandidate,
+    SingleCandidateRequirement,
+)
+from pip._internal.utils.typing import MYPY_CHECK_RUNNING
+
+if MYPY_CHECK_RUNNING:
+    from typing import Any, List, Union, Sized
+    from pip._internal.index.package_finder import PackageFinder
+    from pip._internal.models.candidate import InstallationCandidate
+    from pip._internal.operations.prepare import RequirementPreparer
+
+    Requirement = Union[PEP440Requirement, SingleCandidateRequirement]
+    Candidate = Union[InstallationCandidate, DirectCandidate, ExtrasCandidate]
+    Dependency = Union[Requirement, Candidate]
+
+
+class Provider(object):
+    def __init__(self, finder, preparer):
+        # type: (PackageFinder, RequirementPreparer) -> None
+        super(Provider, self).__init__()
+        self.finder = finder
+        self.preparer = preparer
+
+    def identify(self, dependency):
+        # type: (Dependency) -> str
+        return canonicalize_name(dependency.name)
+
+    def get_preference(self, resolution, candidates, information):
+        # type: (Any, Sized[Candidate], Any) -> int
+        return len(candidates)
+
+    def find_matches(self, req):
+        # type: (Requirement) -> List[Candidate]
+        if isinstance(req, SingleCandidateRequirement):
+            return [req.candidate]
+
+        if req.url:
+            candidates = [DirectCandidate(req.name, req.url)]
+        else:
+            candidates = self.finder.find_all_candidates(req.name)
```

Comment:
This should probably use `find_best_candidate` - we don't actually want the finder's choice of "best", but we do want the list of applicable candidates, in preference order, that `find_best_candidate` calculates, as that drops (for example) incompatible wheels. It also implements the rules that we prefer wheels over sdists, and "more specific" wheels over "less specific" ones.

Reply:
If I read it correctly, `find_best_candidate()` does not actually order the candidates by preference, so I added a method on the evaluator to do that 😐

Reply:
Yeah, I think I missed that in my prototype. Good catch.

Reply:
It does -- https://github.com/pypa/pip/pull/7332. You can use `BestCandidateResult.applicable_candidates`.